core/src/nodes/handled_node_tasks.rs

```diff
@@ -450,3 +451,380 @@ where
+            assert_matches!(events[0], (InToExtMessage::TaskClosed{..}, TaskId(345)));
+        }
+
+        #[test]
+        fn task_exits_when_node_is_done() {
+            let mut rt = Runtime::new().unwrap();
+            let fut = {
+                let peer_id = PublicKey::Rsa((0 .. 2048).map(|_| -> u8 { random() }).collect()).into_peer_id();
+                let mut muxer = DummyMuxer::new();
+                muxer.set_inbound_connection_state(DummyConnectionState::Closed);
+                muxer.set_outbound_connection_state(DummyConnectionState::Closed);
+                future::ok((peer_id, muxer))
+            };
+            let (node_task, tx, rx) = TestBuilder::new()
+                .with_inner_fut(fut)
+                .with_task_id(345)
+                .node_task();
+
+            // Even though we set up the muxer outbound state to be `Closed` we
+            // still need to open a substream or the outbound state will never
+            // be checked.
+            // We do not have a HandledNode yet, so we can't simply call
+            // `open_substream`. Instead we send a message to the NodeTask,
+            // which will be buffered until the inner future resolves, then
+            // it'll call `inject_event` on the handler. In the test Handler,
+            // inject_event will set the next state so that it yields an
+            // OutboundSubstreamRequest.
+            // Back up in the HandledNode, at the next iteration we'll
+            // open_substream() and iterate again. This time, node.poll() will
+            // poll the muxer inbound (closed) and also outbound (because now
+            // there's an entry in the outbound_streams) which will be Closed
+            // (because we set up the muxer state so) and thus yield
+            // Async::Ready(None) which in turn makes the NodeStream yield an
+            // Async::Ready(OutboundClosed) to the HandledNode.
+            // Now we're at the point where poll_inbound, poll_outbound and
+            // address are all skipped and there is nothing left to do: we yield
+            // Async::Ready(None) from the NodeStream. In the HandledNode,
+            // Async::Ready(None) triggers a shutdown of the Handler so that it
+            // also yields Async::Ready(None). Finally, the NodeTask gets a
+            // Async::Ready(None) and sends a TaskClosed and returns
+            // Async::Ready(()). Qed.
+
+            let create_outbound_substream_event = InEvent::Substream(Some(135));
+            tx.unbounded_send(create_outbound_substream_event).expect("send msg works");
+            rt.spawn(node_task);
+            let events = rt.block_on(rx.collect()).expect("rx failed");
+
+            assert_eq!(events.len(), 2);
+            assert_matches!(events[0].0, InToExtMessage::NodeReached(PeerId{..}));
+            assert_matches!(events[1].0, InToExtMessage::TaskClosed(Ok(()), _));
+        }
+    }
+    mod handled_node_tasks {
```

Comment:
Why multiple submodules instead of putting everything in the same?

Reply:
It was a way to let me run the just the relevant tests while writing them, `cargo t handled_node_tasks` and in the end it seemed like a decent way to group tests. Not to your liking I take it?